File: Python/6py 2/rev_rot.py
"""
|--------------------------------------------------------------------------
| Reverse or rotate?
|--------------------------------------------------------------------------
|
"""

def rev_rot(strng, sz):
   if sz <= 0 or strng == "" or sz > len(strng): return ""
   res = ""
   for i in range(0, len(strng), sz):
      chunk = strng[i:i + sz]
      if len(chunk) == sz:
         chunk_sum = sum(int(digit) for digit in chunk)
         res += chunk[::-1] if chunk_sum % 2 == 0 else chunk[1:] + chunk[0]
   return res


def rev_rot(strng, sz):
   chunks = [strng[i*sz:(i+1)*sz] for i in range(len(strng) // sz)] if sz > 0 else []
   return ''.join([[c[::-1], c[1:] + c[0]][sum([int(x) ** 2 for x in c]) & 1] for c in chunks])


def rev_rot(strng, sz):
   if sz<=0 or len(strng)==0 or len(strng)<sz: return ""
   arr= []
   while len(strng)>=sz:
      arr.append(strng[:sz])
      strng= strng[sz:]
   return "".join( x[1:]+x[0] if sum(int(d)**3 for d in x)%2 else x[::-1] for x in arr)


print(rev_rot("1234", 0), "")
print(rev_rot("", 0), "")
print(rev_rot("1234", 5), "")
print(rev_rot("733049910872815764", 5), "330479108928157")

File: JavaScript/6js 3/revrot.js
/*
|--------------------------------------------------------------------------
| Reverse or rotate?
|--------------------------------------------------------------------------
|
*/

function revrot(str, sz) {
   if (sz <= 0 || str === "" || sz > str.length) return "";
   let res = "";
   for (let i = 0; i < str.length; i += sz) {
      let chunk = str.slice(i, i + sz);
      if (chunk.length === sz) {
         let sum = chunk.split("").reduce((acc, curr) => acc + parseInt(curr), 0);
         res += sum % 2 === 0 ? chunk.split("").reverse().join("") : chunk.slice(1) + chunk[0];
      }
   }
   return res;
}


function revrot(str, sz) {
   return (str.match(new RegExp(`.{${sz}}`, `g`)) || [])
      .map((val) =>
         val.replace(/[02468]/g, ``).length % 2
            ? val.replace(/(.)(.+)/, `$2$1`)
            : [...val].reverse().join(``)
      )
      .join(``);
}


function revrot(str, sz) {
   if (sz <= 0 || str.length < sz) return "";
   const chunks = [];
   while (str.length >= sz) {
      chunks.push(str.slice(0, sz));
      str = str.slice(sz);
   }
   return chunks
      .map((item) => {
         const itemArr = item.split("");
         const kubs = itemArr.reduce((res, num) => res + Math.pow(num, 3), 0);
         return (kubs % 2 ? [...itemArr.slice(1), itemArr.shift()] : itemArr.reverse()).join("");
      })
      .join("");
}


console.log(revrot("1234", 0), "");
console.log(revrot("", 0), "");
console.log(revrot("1234", 5), "");
console.log(revrot("733049910872815764", 5), "330479108928157");